src/hooks/useImageInputCompression.ts

```diff
@@ -0,0 +1,75 @@
+import { ChangeEvent, useEffect, useMemo, useState } from "react";
+import { FieldError, useForm } from "react-hook-form";
+import { useAsyncFn } from "react-use";
+
+import {
+  GIF_RESIZER_URL,
+  MAX_SELECTABLE_IMAGE_FILE_SIZE_BYTES,
+  MAX_SELECTABLE_IMAGE_FILE_SIZE_MB,
+} from "settings";
+
+import { useTryCompressImage } from "hooks/useTryCompressImage";
+
+export const imageCompressionErrorMessage =
+  "An error occurred while compressing the image.";
+
+export const tooLargeFileSelectedErrorMessage = `File size limit is ${MAX_SELECTABLE_IMAGE_FILE_SIZE_MB}MB. You can shrink images at ${GIF_RESIZER_URL}`;
+
+export const useImageInputCompression = (
+  register: ReturnType<typeof useForm>["register"],
+  externalErrorMessage: FieldError["message"] | undefined,
+  fileInputName: string
+) => {
+  useEffect(() => {
+    register(fileInputName);
+  }, [fileInputName, register]);
+
+  const [isTooLargeFileError, setIsTooLargeFileError] = useState(false);
+  const tryCompress = useTryCompressImage();
+
+  const [
+    { loading, error: isCompressionError },
+    handleFileInputChange,
+  ] = useAsyncFn(
+    async (
+      event: ChangeEvent<HTMLInputElement>
+    ): Promise<[string | undefined, File | undefined]> => {
+      const files = event.target.files;
+      const file = files?.[0];
+
+      if (!files || !file) return [undefined, undefined];
+
+      if (file.size >= MAX_SELECTABLE_IMAGE_FILE_SIZE_BYTES) {
+        setIsTooLargeFileError(true);
+        return [undefined, undefined];
+      } else setIsTooLargeFileError(false);
+
+      const compressedFile = await tryCompress(file);
+      const url = URL.createObjectURL(file);
+
+      return [url, compressedFile];
```

Comment:
Objects introduce unnecessary renames as we cannot properly name variables in an object for every possible hook usage. And tuples are widely used in React and it's ecosystem: `useState`, `useReducer` etc.
The good example of this is our `useShowHide` hook where in the vast majority of it's usages it's either: 
```typescript
  const {
    isShown: isJoinMessageVisible,
    show: showJoinMessage,
    hide: hideJoinMessage,
  } = useShowHide(false);
```
or

```typescript
  const {
    isShown: showAdvancedSettings,
    toggle: toggleShowAdvancedSettings,
  } = useShowHide(false);
```

In the first example, it may be better to extract it to a separate hook, like `useBooleanState` (like this [one](https://chakra-ui.com/docs/hooks/use-boolean)) and in the second - to use `useToggle` hook from `react-use` lib.

It's good to return an object when there are a lot of parameters returned (because indeed tuples introduce confusion as it's hard to memorize the exact order of, say, 4 returned values) but in case of two values - I believe it's more beneficial to use a tuple. What do you think?

Reply:
although, I wouldn't say that I totally support your approach, and it'd be easier for me to have an object, rather than an array (don't see any problems in renaming), but I've just asked @mike-lvov and he is on your side in this case, when we have URL and file of the same entity. So agreed, let's stick with your approach here.